### PR TITLE
ARC tunable callback, arc_flush, dbuf cache kstats

### PR DESCRIPTION
**Pull request by dinatale2**
### Description
When changing the following ARC tunings immediately apply the
change to the ARC instead of deferring it until the arc_reclaim
thread runs.  In order to keep the logic simple several types
were changed to unsigned longs.

  zfs_arc_min
  zfs_arc_max
  zfs_arc_meta_min
  zfs_arc_meta_limit
  zfs_arc_meta_limit_percent
  zfs_arc_dnode_limit
  zfs_arc_dnode_limit_percent
  zfs_arc_grow_retry
  zfs_arc_shrink_shift
  zfs_arc_p_min_shift
  zfs_arc_min_prefetch_lifespan
  zfs_arc_lotsfree_percent
  zfs_arc_sys_free

Introduce arc_flush module option.

Introduce relevant kstats for the dbuf cache to help
gain insight into how the dbuf cache is functioning.
This kstat can be useful to debug potential caching
issues or help determine if a new change impacts dbuf
cache behavior.

### How Has This Been Tested?

Locally on a VM.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux code style requirements.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by richardelling** [comment]
I'll look into this in more depth soon. But first, a system-level question: suppose arc_flush_impl() takes a long time, say hours, to complete. What happens to the rest of the system? For background, it is quite possible to see tens of minutes required to flush the contents of a pool upon export for a system with lots of RAM.

**Comment by chrisrd** [comment]
How about documenting `zfs_arc_flush` in `man/man5/zfs-module-parameters.5`?

...or not? We have a lot of undocumented params (e.g. I was just looking at `dbuf_cache_*`, also undocumented). Is the general view that we'd (eventually) like all params documented, or are some params deliberately undocumented because "here be dragons" and they're really only for highly experienced users who are reading and understanding the source - basically hidden params (although they're not very well hidden given the existence of `/sys/module/zfs/parameters/*`)?

**Comment by behlendorf** [comment]
@richardelling good question.  This patch doesn't change any of the existing behavior of `arc_flush_impl()` which has always been best effort.  Any buffers which are busy or otherwise pinned will be skipped.  It's primarily intended to make it possible to write test cases which verify the ARC is functioning correctly and to help facilitate performance analysis.  

As for pool taking tens of minutes to export I'd be interested to see where all that time is being spent.  That's much longer than I would have expected unless you have a ton of dirty data which needs to be written.

> How about documenting zfs_arc_flush in man/man5/zfs-module-parameters.5?

Let's definitely do this if we stick with `zfs_arc_flush` as being the interface for this.  Another option would be to extend `zinject -a`.  This has the advantage of making it clear it's for debugging, getting all the gross text parsing out of the kernel and it makes it easy to call from the ZTS.  The downside if we'll probably need to use some of the pad space in `zinject_record_t` and will have the usual user/kernel compatibility concerns.

> We have a lot of undocumented params, ... Is the general view that we'd (eventually) like all params documented

Yes.  We'd like to document everything exposed as a module option.  We can, and should, include warnings as appropriate for the more dangerous knobs.

**Comment by dinatale2** [comment]
Refreshed. Accidentally used `dbuf_cache_hiwater_pct` instead of `dbuf_cache_lowater_pct` in `dbuf_cache_lowater_bytes()`.

One other question for everyone. Are there any other dbuf cache stats that I should be tracking?

**Comment by richardelling** [comment]
@dinatale2 I plan to add a few dbufstats, but I'm not yet happy with them and they can be added later. I'd rather not delay this work.

File: module/zfs/arc.c [hunk 1]
```diff
@@ -4606,9 +4619,64 @@ arc_adjust(void)
 	return (total_evicted);
 }
 
+static void
+arc_flush_impl(flush_state_t *fs)
+{
+	uint64_t guid = fs->pool_guid;
+	boolean_t retry = fs->retry;
+
+	if (fs->arc_mru == B_TRUE && fs->data == B_TRUE)
```
[the remark this answers is not in the document]
Comment (dinatale2) [reply module/zfs/arc.c]: @richardelling This is what I came up with. I'm not sure I like that the names of the member variables in the `flust_state_t` struct have to be consistent with the names of the `arc_state_t` globals. Thoughts?

```
#define ARC_FLUSH_METADATA(fs, guid, list, retry) {                     \
        if(fs->list == B_TRUE && fs->metadata == B_TRUE)                \
                (void) arc_flush_state(list, guid, ARC_BUFC_METADATA, retry); \
}

#define ARC_FLUSH_DATA(fs, guid, list, retry) {                 \
        if(fs->list == B_TRUE && fs->data == B_TRUE)            \
                (void) arc_flush_state(list, guid, ARC_BUFC_DATA, retry); \
}
```

**Comment by dinatale2** [comment]
@richardelling I went ahead and refreshed this PR with more stats. I put in a separate `hash_insert_race` stat so we can keep track of the case when a hash insert occurs with the same dbuf.

@grwilson This version includes the histograms you were recommending at the conference. Would you mind giving me your thoughts on the stats we are tracking?

File: module/zfs/arc.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (richardelling) [reply module/zfs/arc.c]: still too complicated :-)  I'm thinking something like
```
if (fs->arc_mru == B_TRUE) {
  if (fs->metadata == B_TRUE) {
    (void) arc_flush_state(arc_mru, guid, ARC_BUFC_METADATA, retry);
  }  /* else? */
  if (fs->data == B_TRUE) {
    (void) arc_flush_state(arc_mfu, guid, ARC_BUFC_DATA, retry);
  }
}
...
```
admittedly, still ugly... but not much we can do with so many binary decisions

File: module/zfs/dbuf.c [hunk 2]
```diff
@@ -3989,6 +4162,22 @@ EXPORT_SYMBOL(dmu_buf_set_user_ie);
 EXPORT_SYMBOL(dmu_buf_get_user);
 EXPORT_SYMBOL(dmu_buf_get_blkptr);
 
+#include <linux/mod_compat.h>
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: We're now using this header in a couple files, let's move the include in to the __KERNEL section of `zfs_context.h`.

File: module/zfs/dbuf.c [hunk 3]
```diff
@@ -619,7 +720,24 @@ dbuf_evict_notify(void)
...
+		dbuf_evict_one();
 
+		size = refcount_count(&dbuf_cache_size);
+		if (size < prev_size)
+			retries = 0;
+		else
+			retries++;
+		prev_size = size;
+	}
+}
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: We shouldn't add an entirely new module option for this, particularly without a clear use case in mind.  We have more than we need already.  I could potentially see an argument for adding a `dbuf_cache` token to `arc_flush`.

File: module/zfs/arc.c [hunk 4]
```diff
@@ -8958,104 +9025,280 @@ EXPORT_SYMBOL(arc_getbuf_func);
...
+	{ TOKEN_TYPE_ALL,		"all-types" },
+	{ TOKEN_POOL_NAME,		"pool=%s" },
+	{ TOKEN_ALL,			"all" },
+	{ TOKEN_LAST,			NULL },
+};
 
-module_param(zfs_arc_meta_strategy, int, 0644);
-MODULE_PARM_DESC(zfs_arc_meta_strategy, "Meta reclaim strategy");
+static int
+parse_arc_flush(char *option, int token, substring_t *args, flush_state_t *fs)
```
Comment (behlendorf) [inline module/zfs/arc.c]: Doing this kind of parsing in the kernel really isn't ideal, although it was expedient at the time.  We should pull this up in to `zinject(8)` and pass the resulting flags in the `zinject_record_t`.  We're only talking about 8 bits and we could claim all 32 from the `zi_pad`.

File: cmd/dbufstat/dbufstat.py [hunk 5]
```diff
@@ -70,9 +70,10 @@
     "meta":       [4,    -1, "is this buffer metadata?"],
     "state":      [5,    -1, "state of buffer (read, cached, etc)"],
     "dbholds":    [7,  1000, "number of holds on buffer"],
+    "dbc":        [3,    -1, "in dbuf cache?"],
     "list":       [4,    -1, "which ARC list contains this buffer"],
     "atype":      [7,    -1, "ARC header type (data or metadata)"],
-    "flags":      [8,    -1, "ARC read flags"],
+    "flags":      [9,    -1, "ARC read flags"],
```
[the remark this answers is not in the document]
Comment (dinatale2) [reply cmd/dbufstat/dbufstat.py]: Yes it does.

File: module/zfs/dbuf.c (hunk 2 above)
Comment (dinatale2) [reply module/zfs/dbuf.c]: Will this still need to happen if I move the dbuf cache flushing to the arc flush function?

Comment (behlendorf) [reply module/zfs/dbuf.c]: No, the header's just needed for `module_param_call()`.

**Comment by dinatale2** [comment]
Closing this PR because it needs to be split up into two separate PRs. I'll open a PR for the dbuf stats in a few moments. I'll need to find time to revisit the ARC flushing.

**Pull request closed without merging**